helm-charts/images/gcp-filestore-backups/gcp-filestore-backups.py

```diff
@@ -0,0 +1,134 @@
+import argparse
+import json
+import subprocess
+import time
+from datetime import datetime, timedelta
+
+import jmespath
+
+
+def main(args):
+    # Get a JSON object of the filestore backups. We filter by project and region
+    # in the gcloud command.
+    filestore_backups = subprocess.check_output(
+        [
+            "gcloud",
+            "filestore",
+            "backups",
+            "list",
+            "--format=json",
+            f"--project={args.project}",
+            f"--region={args.region}",
+        ],
+        text=True,
+    )
+    filestore_backups = json.loads(filestore_backups)
+
+    # Filter returned backups by filestore and share names
+    filestore_backups = jmespath.search(
+        f"[?sourceFileShare == '{args.filestore_share_name}' && contains(sourceInstance, '{args.filestore_name}')]",
+        filestore_backups,
+    )
+
+    # Parse `createTime` to a datetime object for comparison
+    filestore_backups = [
+        {
+            k: (
+                datetime.strptime(v.split(".")[0], "%Y-%m-%dT%H:%M:%S")
+                if k == "createTime"
+                else v
+            )
+            for k, v in backup.items()
+        }
+        for backup in filestore_backups
+    ]
+
+    # Generate a list of filestore backups that are less than 24 hours old
+    recent_filestore_backups = [
+        backup
+        for backup in filestore_backups
+        if datetime.now() - backup["createTime"] < timedelta(days=1)
+    ]
+
+    # Generate a list of filestore backups that are older than our set retention period
+    old_filestore_backups = [
+        backup
+        for backup in filestore_backups
+        if datetime.now() - backup["createTime"] > timedelta(days=args.retention_days)
+    ]
+
+    if len(recent_filestore_backups) == 0:
+        print(
+            f"There have been no recent backups of the filestore for project {args.project}. Creating a backup now..."
+        )
+
+        subprocess.check_call(
+            [
+                "gcloud",
+                "filestore",
+                "backups",
+                "create",
+                f"{args.filestore_name}-{args.filestore_share_name}-backup-{datetime.now().strftime('%Y-%m-%d')}",
+                f"--file-share={args.filestore_share_name}",
+                f"--instance={args.filestore_name}",
+                f"--instance-location={args.region}-b",
+                f"--region={args.region}",
+                # This operation can take a long time to complete and will only take
+                # longer as filestores grow, hence we use the `--async` flag to
+                # return immediately, without waiting for the operation in progress
+                # to complete. Given that we only expect to be creating a backup
+                # once a day, this feels safe enough to try for now.
+                "--async",
```

Comment:
I think calling it `--async` is fine, as long as our logic to detect 'do we need to create another backup right now?' takes into account incomplete backups. We want to avoid running into the [rate limit](https://cloud.google.com/filestore/docs/limits) and also avoid endlessly creating a ton of backups (as that can cost money).

Reply:
I checked and it looks like this is handled on GCP's side. I ran the script twice in a row, and the returned JSON was already populated with the backup just created, even though it's state was "FINALIZING". So I didn't need to make any tweaks to the script here.

```
$ python gcp-filestore-backups.py pilot-hubs-homedirs two-eye-two-see us-central1-b
There have been no recent backups of the filestore for project two-eye-two-see. Creating a backup now...
Create request issued
Check operation [projects/two-eye-two-see/locations/us-central1/operations/operation-1720791941383-61d0d18660679-d749fa00-3bdf9958] for status. 
No outdated backups found.
...

$ python gcp-filestore-backups.py pilot-hubs-homedirs two-eye-two-see us-central1-b
[
  {
    "capacityGb": "5120",
    "createTime": "2024-07-12T13:45:41.576982250Z",
    "name": "projects/two-eye-two-see/locations/us-central1/backups/pilot-hubs-homedirs-homes-backup-2024-07-12",
    "sourceFileShare": "homes",
    "sourceInstance": "projects/two-eye-two-see/locations/us-central1-b/instances/pilot-hubs-homedirs",
    "sourceInstanceTier": "BASIC_HDD",
    "state": "FINALIZING"
  }
]
Recent backup found.
No outdated backups found.
...
```